n_queen_problem/NQueenProblem.java

```diff
@@ -1,64 +1,171 @@
-/*
- * Following code is the implementation of the N-queen problem.
- * The solution is to place the queens on board such that no queen is under attack by another queen.
- * There can be a number of possible solutions for a given length of board.
- * This implementation prints only one valid solution, it can be extended to print all possible valid solutions.
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.List;
+
+
+/**
+ * <p>This class solves the N Queens Problem.</p>
+ *
+ * <p>Once you create an object with the desired board size, you can get an iterator that lazily will create each
+ * possible solution is there's any. The code will create all solutions, not jut unique ones</p>
+ *
+ * <p>This way you have flexibility on how to use solutions, you can get only one or all.</p>
+ *
+ * @see <a href="https://en.wikipedia.org/wiki/Eight_queens_puzzle">N Queens Problem</a>
  */
+public class NQueenProblem implements Iterable<NQueenProblem.Solution> {
+
+    private final int size;
 
-import java.util.Arrays;
+    /**
+     * Creates an NQueens object to solve the N Quees Problems on board of size n
+     *
+     * @param n
+     */
+    public NQueenProblem(final int n) {
+        this.size = n;
+    }
 
-public class NQueenProblem {
+    /**
+     * Returns an Iterator over the solutions for this N Queens Problem.
+     */
+    @Override
+    public Iterator<Solution> iterator() {
+        return new NQueenIterator();
+    }
 
-    private static boolean isQueenSafe(int[][] board, int row, int column) {
-        int i;
-        int j;
-        int boardLength = board[0].length;
-        for (i = 0; i < column; ++i)
-            if (board[row][i] == 1)
-                return false;  // If there's another Queen present in the same row.
+    /**
+     * <p>This class represents a Solution for the N Queen problem.</p>
+     *
+     * <p>Internally a solution is encoded as a list of integers. If in the index i there's integer j
+     * it means that one of the queens should be placed on the square (i, j)</p>
+     *
+     */
+    public static class Solution {
 
-        for (i = row, j = column; i >= 0 && j >= 0; --i, --j)
-            if (board[i][j] == 1)
-                return false;  // If there's another Queen present in the upper diagonal.
+        private final List<Integer> board;
+        private static char QUEEN = 'Q';
+        private static char EMPTY_SQUARE = '.';
+        private static String NEW_LINE = String.format("%n");
```

Comment:
Can you please explain about `"%n"` here? Can't we simply make it equal to `"\n"`?

Reply:
"\n" is platform dependent because it works in Windows systems but not on *nix OS's. "%n" was created to address that. I think the name of the constant explain what it does and if you want the specific you can check the API. But I agree it doesn't hurt a little comment